### PR TITLE
Proj 200 prior py ivl feature

### PR DESCRIPTION
**Pull request by saipraveen18**
### Description

This PR allows consumers to plan shop in previous year. 
Display user information in View sep history admin action.

```
bundle exec rake parallel:spec[4]
bundle exec cucumber
```

### CODE REVIEW

File: app/views/sep_history/_sep_history_display.html.erb [hunk 1]
```diff
@@ -0,0 +1,90 @@
+<% @family.special_enrollment_periods.to_a.reverse_each do |f| %>
```
Comment (saimekala07) [inline app/views/sep_history/_sep_history_display.html.erb:1]: Do we need to_a, can we sort_by created_at date

File: app/models/special_enrollment_period.rb [hunk 2]
```diff
@@ -238,6 +251,17 @@ def set_submitted_at
...
 
+  def set_coverage_renewal_flag
+    prior_py_ivl_sep = EnrollRegistry.feature_enabled?(:prior_plan_year_ivl_sep)
+    return if coverage_renewal_flag == false || prior_py_ivl_sep == false
+
+    self.assign_attributes({coverage_renewal_flag: prior_py_ivl_sep})
+  end
+
+  def set_user_id
+    self.user_id = SAVEUSER[:current_user_id]
```
Comment (saimekala07) [inline app/models/special_enrollment_period.rb:262]: Looks like user_id is not saving, can you check after SEP record in DB.

File: app/views/sep_history/_sep_history_display.html.erb [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+          </div>
+          <div class="col-md-6">
+            <div class="col-md-6 font-weight-bold"><%= l10n("hbx_profiles.sep_event_date") %></div>
+            <div class="col-md-6"><%= f.qle_on %></div>
+          </div>
+        </div>
+        <div class="row mt-1">
+          <div class="col-md-6">
+            <div class="col-md-6 font-weight-bold"><%= l10n("hbx_profiles.sep_creator") %></div>
+            <div class="col-md-6"><%= f.user_email.present? ? f.user_email : 'N/A' %></div>
```
Comment (saimekala07) [inline app/views/sep_history/_sep_history_display.html.erb:46]: Screen Shot 2021-05-26 at 2 20 57 PM

By default it is displaying current user whoever logs.

File: app/models/special_enrollment_period.rb [hunk 4]
```diff
@@ -68,6 +68,12 @@ class SpecialEnrollmentPeriod
   # ADMIN FLAG
   field :admin_flag, type:Boolean
 
+  # USER ID
+  field :user_id, type: BSON::ObjectId
+
+  #Renew coverage flag
+  field :coverage_renewal_flag, type: Boolean
```
Comment (saimekala07) [inline app/models/special_enrollment_period.rb:75]: `can_renew_coverage` by default this value needs to be true

File: app/helpers/hbx_admin_helper.rb [hunk 5]
```diff
@@ -63,4 +63,14 @@ def find_enrollment(hbx_id)
...
     family.active_household.latest_active_tax_household_with_year(TimeKeeper.date_of_record.year) ? 'Yes' : 'No'
   end
+
+  def prior_py_sep?(family, effective_date, market)
+    person = family.primary_person
+    ivl_prior_coverage_period = HbxProfile.current_hbx.benefit_sponsorship.current_benefit_coverage_period
+    if market == 'individual'
+      !ivl_prior_coverage_period&.contains?(effective_date)
+    else
+      person.active_employee_roles.none?{|e| e.census_employee&.active_benefit_package&.benefit_application&.effective_period&.cover?(effective_date)}
```
Comment (saimekala07) [inline app/helpers/hbx_admin_helper.rb:73]: Is this change needed for IVL prior plan year SEP

File: app/controllers/concerns/sep_all.rb [hunk 6]
```diff
@@ -177,6 +178,7 @@ def createSep
     special_enrollment_period.optional_effective_on = date_arr if date_arr.length > 0
     special_enrollment_period.market_kind = qle.market_kind == "individual" ? "ivl" : qle.market_kind
     special_enrollment_period.admin_flag = true
+    special_enrollment_period.coverage_renewal_flag = sep_params[:coverage_renewal_flag].present? ? sep_params[:coverage_renewal_flag] : true
```
Comment (saimekala07) [inline app/controllers/concerns/sep_all.rb:181]: Set this value-based user selection.

File: app/controllers/exchanges/hbx_profiles_controller.rb [hunk 7]
```diff
@@ -907,7 +907,7 @@ def create_ba_params
   def sep_params
     params.except(:utf8, :commit).permit(:market_kind, :person, :firstName, :lastName, :family_actions_id,
                                          :effective_on_kind, :qle_id, :event_date, :effective_on_date, :csl_num,
-                                         :start_on, :end_on, :next_poss_effective_date, :option1_date, :option2_date, :option3_date, :admin_comment)
+                                         :start_on, :end_on, :next_poss_effective_date, :option1_date, :option2_date, :option3_date, :admin_comment, :coverage_renewal_flag, :prior_year_sep_date)
```
Comment (saimekala07) [inline app/controllers/exchanges/hbx_profiles_controller.rb:910]: Do we need `prior_year_sep_date` param

File: app/views/sep/approval/_admin_detail.html.erb [hunk 8]
```diff
@@ -83,18 +84,60 @@
 						    <%= text_field_tag :option3_date, "", :'data-date-max'=> sep_optional_date(@family, 'max'), :'data-date-min'=> sep_optional_date(@family, 'min'),class: "floatlabel required date-picker form-control", placeholder: "DATE 3 OPTION" %>
 						</div>
 					</div>
+          <% if ::EnrollRegistry.feature_enabled?(:prior_plan_year_ivl_sep) %>
+            <div class='row row-form-wrapper no-buffer'>
+              <div class="col-md-12 col-sm-12 col-xs-12 form-group form-group-lg no-pd">
+                <%= check_box_tag :coverage_renewal_flag, 'false' , false, id: "coverage_renewal_flag", class: 'coverage_renewal_checkbox' %>
```
Comment (saimekala07) [inline app/views/sep/approval/_admin_detail.html.erb:90]: coverage_renewal_flag need to display based on some conditions,

For example if user reporting sep today, we don't need to display this checkbox becoz their is nothing to renew now.

If we are in OE period or if admin reporting sep in prior plan year we need to display this checkbox.

By default this checkbox need to be true.

File: app/domain/operations/product_selection_effects/dchbx_product_selection_effects.rb [hunk 9]
```diff
@@ -50,6 +54,30 @@ def renew_ivl_if_is_open_enrollment(enrollment)
...
+        return nil if enrollment.is_shop?
+        @benefit_coverage_periods = fetch_bcp_gt_enr_effective_year(enrollment)
+        return nil if @benefit_coverage_periods.empty?
+        renew_enrollments(enrollment)
+        Success(@enrollment)
+      end
+
+      def renew_enrollments(enrollment)
+        @enrollment = enrollment
+        @benefit_coverage_periods.each do |bcp|
```
Comment (saimekala07) [inline app/domain/operations/product_selection_effects/dchbx_product_selection_effects.rb:69]: https://ideacrew.slack.com/archives/G01DT9SNJ0Z/p1622056537033600

File: app/controllers/exchanges/hbx_profiles_controller.rb (hunk 7 above)
Comment (saipraveen18) [reply app/controllers/exchanges/hbx_profiles_controller.rb:910]: 👍

File: app/helpers/hbx_admin_helper.rb (hunk 5 above)
Comment (saipraveen18) [reply app/helpers/hbx_admin_helper.rb:73]: This comes into picture when we allow prior py for shop employees, but this should not bother the current IVL flow

File: app/models/special_enrollment_period.rb (hunk 2 above)
Comment (saipraveen18) [reply app/models/special_enrollment_period.rb:262]: Will look into it

File: app/views/sep_history/_sep_history_display.html.erb (hunk 3 above)
Comment (saipraveen18) [reply app/views/sep_history/_sep_history_display.html.erb:46]: I think having email makes more sense to display, or we need to display hbx_id

**Review by saimekala07 — APPROVED** [review]
Allowing admin and user to report SEP(IVL) in the prior year and renew coverage accordingly. Updating SEP history to display with created-by attributes.

**Review by saimekala07 — APPROVED** [review]
Fixing incorrect renewal generation bug and Admin data picker for reported qles
